### PR TITLE
Add VK_KHR_image_format_list

### PR DESCRIPTION
**Pull request by sfricke-samsung**
- VUID-VkImageViewCreateInfo-pNext-01585
- VUID-VkImageViewCreateInfo-pNext-04082
- VUID-VkImageViewCreateInfo-flags-04083

### CODE REVIEW

File: layers/parameter_validation_utils.cpp [hunk 1]
```diff
@@ -1161,6 +1162,28 @@ bool StatelessValidation::manual_PreCallValidateCreateImage(VkDevice device, con
...
+            const uint32_t viewFormatCount = format_list_info->viewFormatCount;
+            if (((image_flags & VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT) == 0) && (viewFormatCount > 1)) {
+                skip |= LogError(device, "VUID-VkImageViewCreateInfo-flags-04083",
+                                 "vkCreateImage(): If the VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT is not set, then "
+                                 "VkImageFormatListCreateInfo::viewFormatCount (%u) must be 0 or 1.",
+                                 viewFormatCount);
+            }
+            // Using the first format, compare the rest of the formats against it that they are compatible
+            for (uint32_t i = 1; i < viewFormatCount; i++) {
+                if (FormatCompatibilityClass(format_list_info->pViewFormats[0]) !=
```
Comment (ncesario-lunarg) [inline layers/parameter_validation_utils.cpp:1177]: Looks like maybe the compatibility check should be between `pCreateInfo->format` and `format_list_info->pViewFormats[i]` over the range `[0, viewFormatCount)`? That's based on what I'm reading in `VUID-VkImageCreateInfo-pNext-04737`, though. I'm not seeing 04082 in the latest published spec (though I do see it in older ones).

File: tests/vklayertests_buffer_image_memory_sampler.cpp [hunk 2]
```diff
@@ -13517,3 +13517,90 @@ TEST_F(VkLayerTest, InvalidShadingRateUsage) {
...
+    }
+
+    ASSERT_NO_FATAL_FAILURE(InitState());
+
+    // Use sampled formats that will always be supported
+    // Last format is not compatible with the rest
+    const VkFormat formats[4] = {VK_FORMAT_R8G8B8A8_UNORM, VK_FORMAT_R8G8B8A8_SNORM, VK_FORMAT_R8G8B8A8_UINT, VK_FORMAT_R8_UNORM};
+    VkImageFormatListCreateInfo formatList = {};
+    formatList.sType = VK_STRUCTURE_TYPE_IMAGE_FORMAT_LIST_CREATE_INFO;
+    formatList.pNext = nullptr;
```
Comment (ncesario-lunarg) [inline tests/vklayertests_buffer_image_memory_sampler.cpp:13540]: I think you can collapse these into a single LvlInitStruct call if you want. Reminds me that I think I still need to set pNext = null in those helpers :grimacing:

File: layers/parameter_validation_utils.cpp [hunk 3]
```diff
@@ -1161,6 +1162,28 @@ bool StatelessValidation::manual_PreCallValidateCreateImage(VkDevice device, con
                              ") must be a multiple of 2.",
                              string_VkFormat(image_format), pCreateInfo->extent.height);
         }
+
+        const auto format_list_info = LvlFindInChain<VkImageFormatListCreateInfo>(pCreateInfo->pNext);
+        if (format_list_info) {
+            const uint32_t viewFormatCount = format_list_info->viewFormatCount;
+            if (((image_flags & VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT) == 0) && (viewFormatCount > 1)) {
+                skip |= LogError(device, "VUID-VkImageViewCreateInfo-flags-04083",
```
Comment (ncesario-lunarg) [inline layers/parameter_validation_utils.cpp:1170]: I also don't see this VUID in the most recent published spec; looks like maybe it was changed to `VUID-VkImageCreateInfo-flags-04738`?

**Review by ncesario-lunarg — COMMENTED** [review]
This mostly LGTM, just some questions about VUIDs on my part.

File: layers/buffer_validation.cpp [hunk 4]
```diff
@@ -5677,6 +5677,24 @@ bool CoreChecks::PreCallValidateCreateImageView(VkDevice device, const VkImageVi
             image_usage = chained_ivuci_struct->usage;
         }
 
+        // If image used VkImageFormatListCreateInfo need to make sure a format from list is used
+        const auto format_list_info = LvlFindInChain<VkImageFormatListCreateInfo>(image_state->createInfo.pNext);
+        if (format_list_info) {
```
[the remark this answers is not in the document]
Comment (sfricke-samsung) [reply layers/buffer_validation.cpp:5682]: good catch, you are right, also 100% should have a test for that, will add

File: layers/parameter_validation_utils.cpp (hunk 3 above)
Comment (sfricke-samsung) [reply layers/parameter_validation_utils.cpp:1170]: it did, its `04737` and `04738` now

I was using the old one, then I realized they should be in `vkCreateImage` and not `vkCreatImageView` and made a spec fix, the spec was fixed and so I just grabbed this branch I had laying around and pushed up... but seem to forgot the simple step of updating the VU strings

File: tests/vklayertests_buffer_image_memory_sampler.cpp (hunk 2 above)
Comment (sfricke-samsung) [reply tests/vklayertests_buffer_image_memory_sampler.cpp:13540]: ya, will fix

also if you have more time sooner, feel free to self-assign https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/2628 as I never got to that as fast as I was expecting I would

File: layers/parameter_validation_utils.cpp (hunk 1 above)
Comment (sfricke-samsung) [reply layers/parameter_validation_utils.cpp:1177]: this was updated when moved to vkImage, will fix as well

**Review by ncesario-lunarg — APPROVED** [review]
LGTM!

**Comment by ncesario-lunarg** [comment]
Running through [internal CI](http://erusea:4040/8868).

**Pull request merged**